design/local-node-agent.md

```diff
@@ -0,0 +1,123 @@
+In this document, we propose a design for a Rook "agent" that will be deployed in the cluster to run on nodes that have a need to perform operations to consume storage from the cluster.
+
+## **Background: Flexvolume Issues**
+In Kubernetes 1.8, the Kubernetes Storage SIG recommends storage providers to implement out-of-tree plugins to provide persistent storage.
+With the full implementation of Container Storage Interface (CSI) several months away from completion, the recommended approach in the meantime is to implement a Flexvolume.
+Deployment of a Flexvolume will be improved in 1.8 to make it less of a manual process (e.g., dynamic discovery), as described in the following pull request: https://github.com/kubernetes/community/pull/833
+
+However, there are still some limitations to the capabilities of a Flexvolume implementation.
+For example, a Flexvolume plugin does not execute in a context that has cluster credentials, so it cannot communicate with the Kubernetes API to perform such operations as creating Custom Resource Definitions (CRDs).
+This document will describe how a Rook local node agent would work with the Flexvolume architecture to provide storage in a Kubernetes cluster, in addition to other responsibilities that must be performed on specific nodes in the cluster.
+
+## **Detailed Design**
+
+### **Responsibilities**
+The Rook agent will have multiple responsibilities beyond just performing storage operations for the Rook Flexvolume driver.
+One can think of the agent as a mini operator that functions at the node level.
+The initial proposed responsibilities of the agent are:
+
+1. Deploy the Rook Flexvolume driver to the `volume-plugin-dir` directory on every node
+1. Perform storage operations on behalf of the Flexvolume driver, such as attaching, detaching, mounting and unmounting Rook cluster storage
+1. Cluster clean up operations, such as forcefully unmapping RBD devices when the Rook cluster is being deleted while there are still pods consuming those volumes
+1. Proxy I/O traffic from kernel modules to user space (e.g. Network Block Device (NBD) kernel module to librbd in userspace)
+
+### **Deployment**
+The Rook operator will deploy the Rook agent to all nodes in the cluster via a Daemonset in the same namespace in which the operator is running.
+It is a permanent (long running) daemon that has a lifecycle tied to the node that it is scheduled on.
+The agent deployment will happen when the operator is first created, in the same flow where the operator is declaring CRDs for clusters, pools, etc.
+This means that the Rook agents are not associated with a specific Rook cluster and that they will be able to handle operations for any Rook cluster instance.
+The Cluster CRD will be updated to allow selectors to control the set of nodes that the agent is scheduled on, but in most cases it is desirable for it to be running on all nodes.
+
```

Comment:
you're right, since the agent is cluster wide, it makes sense for any node selectors to be at the operator level not the cluster CRD level.  i've clarified this.